File: programs/cleaning/cl_04_nass.py
import pandas as pd
import os
import glob

# 1. Define the path to your target NASS directory folder
nass_folder = os.path.join("data", "outcome", "National-Agricultural-Statistics-Service")

# 2. Grab a list of every single CSV file inside that specific directory
csv_files = glob.glob(os.path.join(nass_folder, "*.csv"))

print(f"Found {len(csv_files)} CSV files to process inside the directory.")

# 3. Loop through each discovered CSV file one by one
for file_path in csv_files:
    try:
        # Load the current dataset, forcing the code columns to read as strings
        df = pd.read_csv(file_path, dtype={"state_fips_code": str, "county_code": str})
        
        # Verify if BOTH required column headers are actually present in this file
        if "state_fips_code" in df.columns and "county_code" in df.columns:
            
            # Clean and pad the codes (e.g., turning state '51' and county '43' into standard text sizes)
            state_clean = df["state_fips_code"].str.strip().str.zfill(2)
            county_clean = df["county_code"].str.strip().str.zfill(3)
            
            # Mash the two strings together to append your uniform 5-digit FIPS column
            df["fips_code"] = state_clean + county_clean
            
            # Overwrite the original file on disk with the new column included
            df.to_csv(file_path, index=False)
            print(f" -> Successfully added 'fips_code' to: {os.path.basename(file_path)}")
        else:
            # Safely skip files that don't need this specific column adjustment
            print(f" -> Skipped (Columns not found): {os.path.basename(file_path)}")
            
    except Exception as e:
        # Alert if a file is still locked by Excel or corrupt
        print(f" [!] Error processing {os.path.basename(file_path)}: {str(e)}")

print("\nAll NASS folder datasets have been successfully processed!")

File: programs/cleaning/cl_04_nass.R
library(dplyr)
library(readr)
library(stringr)


#### THIS SCRIPT CAN BE USED FOR OTHER DATASETS AS WELL THAT HAVE STATE AND COUNTY FIPS CODES BUT NOT THE 5 DIGIT ONE ####

# ---------------------------------------------
# 1. Define the folder containing NASS CSV files
# ---------------------------------------------
nass_folder <- file.path("data", "outcome", "National-Agricultural-Statistics-Service")

# ---------------------------------------------
# 2. List all CSV files inside that folder
# ---------------------------------------------
csv_files <- list.files(nass_folder, pattern = "\\.csv$", full.names = TRUE)

cat("Found", length(csv_files), "CSV files to process inside the directory.\n")

# ---------------------------------------------
# 3. Loop through each CSV file
# ---------------------------------------------
for (file_path in csv_files) {
  cat("\nProcessing:", basename(file_path), "\n")
  
  # Try reading the file safely
  try({
    # Read CSV, forcing the two code columns to be character
    df <- read_csv(
      file_path,
      col_types = cols(
        state_fips_code = col_character(),
        county_code = col_character(),
        .default = col_guess()
      )
    )
    
    # ---------------------------------------------
    # 4. Check if both required columns exist
    # ---------------------------------------------
    if (all(c("state_fips_code", "county_code") %in% names(df))) {
      
      # ---------------------------------------------
      # 5. Clean and pad the codes
      #     - state_fips_code → 2 digits
      #     - county_code → 3 digits
      # ---------------------------------------------
      df <- df %>%
        mutate(
          state_clean  = str_pad(str_trim(state_fips_code), width = 2, side = "left", pad = "0"),
          county_clean = str_pad(str_trim(county_code), width = 3, side = "left", pad = "0"),
          
          # ---------------------------------------------
          # 6. Create the 5-digit FIPS code
          # ---------------------------------------------
          fips_code = paste0(state_clean, county_clean)
        ) %>%
        select(-state_clean, -county_clean)  # remove helper columns
      
      # ---------------------------------------------
      # 7. Write updated CSV back to disk
      # ---------------------------------------------
      write_csv(df, file_path)
      
      cat(" -> Successfully added 'fips_code' to:", basename(file_path), "\n")
      
    } else {
      # Skip files missing required columns
      cat(" -> Skipped (Columns not found):", basename(file_path), "\n")
    }
    
  }, silent = TRUE)
}

cat("\nAll NASS folder datasets have been successfully processed!\n")
